### PR TITLE
Add support for Deneb

### LINKED ISSUE #3625: EIP-4844 Tracking
## Description

Thanks a lot @dknopik and @mariusVanDerWijden for kick starting the Lighthouse implementation!

The main Lighthouse EIP-4844 branch will be here: https://github.com/sigp/lighthouse/tree/eip4844

This branch currently has a lot of boiler plate implemented but there is still a lot of functionality missing. The goal in the short term is to at least get this branch far enough along to participate in devnets.  I'll write up issues for the tasks listed below and link them here. All the specs are still under development so the info here will likely change. 

#### Primary
- Block and blob processing
  - We could treat blocks as optimistic until blobs are validated
  - A `BlobHandle` similar to `ExecPayloadHandle` but would be for an RPC request and include a channel receiver from gossip 
  - Verifying kzg commintments against transactions - this is specified in gossip and block processing
- Block production
  - Includes kzg proof construction 
  - KZG implementation status
    - https://github.com/benjaminion/c-kzg - from what I can tell no rust bindings yet
    - https://github.com/ralexstokes/kzg
  - `BlobSidecar` signing
  - Implementation of Beacon API changes to support unsigned and signed blob communication
- Gossip validation
  - blob and block validation are currently independent of each other in gossip 
- Make the execution payload into a `superstruct`
  - Mark and Michael have been working on this in our Capella implementation. We will similarly need this because there are new fee market fields in the execution payload.
#### Secondary
- Blob pruning
- Newtype SSZ derive `#[ssz(transparent)]`

## Outstanding Questions
 - Will blocks and blobs be gossiped and RPC'd together? Currently no on the devnet
 - Should we include `BlobsByRoot` in RPC?
 - Should `BlobsByRange` use signed or unsigned blobs?

## Spec Links

- Meta Spec - https://hackmd.io/@protolambda/eip4844-meta
- Execution API - https://github.com/ethereum/execution-apis/pull/197
- Beacon API - Does not exist yet but we'll need something for block production
- Consensus specs - https://github.com/ethereum/consensus-specs/tree/dev/specs/eip4844
- The [EIP](https://eips.ethereum.org/EIPS/eip-4844) itself
- EIP 4844 Devnet repo - https://github.com/Inphi/eip4844-interop

## Other Useful Links

- Prysm implementation notes https://hackmd.io/hZefl5F5RgWXeibgWflRQQ?view
- Notes on whether to couple block and blob in RPC/gossip https://notes.ethereum.org/RLOGb1hYQ0aWt3hcVgzhgQ?view
- General FAQ https://notes.ethereum.org/@vbuterin/proto_danksharding_faq#What-is-proto-danksharding-aka-EIP-4844
- Readiness checklist https://github.com/ethereum/pm/blob/master/Breakout-Room/4844-readiness-checklist.md

### PR DESCRIPTION
**Pull request by realbigsean**
Resolves #3625

### Devnet V3 related notes [deprecated]

- install lighthouse on this branch by runnnig  `make`
  - Note: docker images for this branch available at `sigp/lighthouse:eip4844-amd64` (portable) and `sigp/lighthouse:eip4844-amd64-modern` (non-portable). Currently working on adding `aarch64` images, not available yet
- run:
  - `lighthouse bn --network eip4844 --execution-jwt <jwtsecret_path> --execution-endpoint http://localhost:8551  --trusted-setup-file <trusted_setup_file_path> --http`
- The `trustued-setup-file` to use: https://github.com/sigp/lighthouse/blob/eip4844/scripts/local_testnet/trusted_setup.txt 

### Devnet V4 related notes
- build the eip-4844 branch of this geth fork: https://github.com/mdehoog/go-ethereum
- `wget https://config.eip4844-devnet-4.ethpandaops.io/el/genesis.json`
- `geth init --datadir geth_data genesis.json`
- run geth
 ```
geth --ipcdisable --http --http.api="eth,web3,net,debug" --syncmode=full --bootnodes "enode://8488a57c323d5e5bef1c40c5cb2330cb0bd75d900afad5e7f50a47ecc1ae7683fb664f509908c6b978633b9af75788e943f26e9b64989504d189ae7b3278ede3@161.35.91.184:30303" --verbosity 4 --datadir geth_data --networkid 4844001004
```
- run lighthouse
```
lighthouse bn --purge-db --network eip4844 --execution-endpoint http://localhost:8551 --execution-jwt geth_data/geth/jwtsecret --debug-level trace --http --disable-optimistic-finalized-sync --libp2p-addresses "/ip4/64.225.74.231/tcp/9000/p2p/16Uiu2HAm4uoHpnkrfLncYRiKjWTtUfKUazH6afQgPtBuNpGDSJ7M"
```

### Local testnet instructions

GNU `grep` and GNU `sed` are required

- install lighthouse
  - `make`
- install lcli 
  - `make install-lcli`
- install this geth commit: https://github.com/mdehoog/go-ethereum/commit/2bd3a3cf9fa7d6a63326daf8c93b34ca5af86663
- update the geth_binary location here: https://github.com/sigp/lighthouse/blob/eip4844/scripts/local_testnet/vars.env#L1
- run `cd scripts/local_testnet`
- run `./start_local_testnet.sh -d debug genesis.json`
- all logs will be in `~/.lighthouse/local-testnet/testnet`
- stop the network with `./stop_local_testnet.sh`

Note: By default, this testnet starts at `Bellatrix`, forks to `Capella` at epoch 1 (slot 32) and forks to `EIP4844` at epoch 2 (slot 64). 

#### Update 12/28/22: "withdrawals" feature is no longer necessary as it's now enabled by default

### CODE REVIEW

**Comment by jimmygchen** [comment]
~~@realbigsean maybe worth adding `json_rpc_snoop` installation to instructions as well:~~

~~- install `json_rpc_snoop`: https://github.com/ethDreamer/json_rpc_snoop~~

~~and your description:~~
~~> this is a proxy between lighthouse and geth that just prints out the json being sent back and forth...it's very useful for debugging CL <> EL compatibility~~

sorry, just realised you've already removed it :)

File: Cargo.toml [hunk 1]
```diff
@@ -63,10 +63,12 @@ members = [
     "consensus/tree_hash_derive",
 
     "crypto/bls",
+    "crypto/kzg",
     "crypto/eth2_hashing",
     "crypto/eth2_key_derivation",
     "crypto/eth2_keystore",
     "crypto/eth2_wallet",
+    "crypto/kzg",
```
Comment (jimmygchen) [inline Cargo.toml:71]: ^ this is added twice

Comment (realbigsean) [reply Cargo.toml:71]: :+1:

**Comment by realbigsean** [comment]
Closing in favor of https://github.com/sigp/lighthouse/pull/4054

**Pull request closed without merging**